mesh/pmesh.cpp

```diff
@@ -969,6 +969,9 @@ void ParMesh::Load(istream &input, int generate_edges, int refine,
       pncmesh->GetConformingSharedStructures(*this);
    }
 
+   // Assume FinalizeTopology() has been called in Loader
+   FinalizeParTopo();
```

Comment:
`ParMesh::Finalize` already calls `FinalizeParTopo`. Is this addition required?

Reply:
Yes, it is, otherwise `sedge_ledge` and similar are not initialized when we get to `GenerateBoundaryElements()`.

Reply:
Maybe I could make it conditional if `refine` or `fix_orientations` are true, because only then the topology is possibly changing in `Finalize`, but I do not know, it might be confusing... 🤔

Reply:
Maybe it would be more clear if inside `ParMesh::Finalize` you just moved `FinalizeParTopo` to be called before `Mesh::Finalize`?

Reply:
I am not sure, the topology might be changing also at other places, there is maybe a good reason for that... 🤔

Reply:
Yeah, it does DOFs reordering with curved meshes, so the edge/face indices are then completely different then and `FinalizeParTopo` must be done after that.

Reply:
> That's true. Another option would be inside of the `ParMesh::GenrateBoundaryElements` overload where these data structures are used. I'm just thinking of how this change could be made more clear/reduce the distance between it and the intended side effects.

I am not sure I follow, do you mean to call `FinalizeParTopo` from there? For some reason it is a public method, so a user may call it any time. So I do not think it is a good idea...

Reply:
I am suggesting to remove the added call here (as others may ask why it's needed after this PR) and replace it with a call at the start of ParMesh::GenerateBoundaryElements(), at line 3209, for example. This way it's clear why you need to call it.

Reply:
Yeah, I would agree when it was a private method how it used to be, but `GenerateBoundaryElements` is public nowadays, so it is dangerous to add side effects to it, when a user may call it any time.

Reply:
How about just a comment in the code? :wink:

Reply:
I think this is an improvement, since it avoids calling `FinalizeParTopo` twice unless absolutely necessary, while also keeping `GenerateBoundaryElements` fine for users to call as a public method:

```
diff --git a/mesh/pmesh.cpp b/mesh/pmesh.cpp
index 999e564c4..9b3db80c2 100644
--- a/mesh/pmesh.cpp
+++ b/mesh/pmesh.cpp
@@ -969,9 +969,6 @@ void ParMesh::Load(istream &input, int generate_edges, int refine,
       pncmesh->GetConformingSharedStructures(*this);
    }

-   // Assume FinalizeTopology() has been called in Loader
-   FinalizeParTopo();
-
    Finalize(refine, fix_orientation);

    EnsureParNodes();
@@ -3206,6 +3203,11 @@ void ParMesh::GenerateBoundaryElements()
       bel_to_edge = NULL;
    }

+   if (shared_edges.Size() != sedge_ledge.Size())
+   {
+      FinalizeParTopo();
+   }
+
    const Array<int> *s2l_face;
    if (Dim == 1) { s2l_face = &svert_lvert; }
    else if (Dim == 2) { s2l_face = &sedge_ledge; }
```

Reply:
I may move `FinalizeParTopo` closer to the source to `LoadSharedEntities` and `ParNCMesh::GetConformingSharedStructures()`, where the structures like `sedge_ledge` are allocated, but not initialized, which is a bad practice by itself. However, I still cannot remove the second call from `ParMesh::Finalize()`, which might be unnecessary sometimes, but it is hard to determine when exactly. In this case of loading from files, it is unnecessary when the topology has not been changed, but there can be other scenarios when the user calls `Finalize` directly (it is a public method) and it is expected to construct these structures (and there is no other way to call `FinalizeParTopo` as it is protected).

Reply:
Indeed it seems like a bad idea that `sface_lface` and `sedge_ledge` are allocated but not filled by `LoadSharedEntities`.

OK, maybe a better suggestion is to follow the code in `ParMesh::MarkTetMeshForRefinement`. Just avoid calling `FinalizeParTopo` altogether and use the `shared_edges`, `shared_trias`, and `shared_quads` members in your `GenerateBoundaryElements` instead. See this block:

```
Array<int> sedge_ord(shared_edges.Size());
Array<Pair<int,int> > sedge_ord_map(shared_edges.Size());
for (int k = 0; k < shared_edges.Size(); k++)
{
   // sedge_ledge may be undefined -- use shared_edges and v_to_v instead
   const int sedge = group_sedge.GetJ()[k];
   const int *v = shared_edges[sedge]->GetVertices();
   sedge_ord[k] = order[v_to_v(v[0], v[1])];
}
```

This might be the safest (and most optimal in terms of not recomputing quantities) route.

Reply:
Yes, that is also an option. However, I have checked other constructors and they also initialize `sedge_ledge` and I do not see any other ways how to manipulate these structures, so it is unnecessary in `Finalize` to set them and it is rather a hidden side effect. It should be safe and a good practice to move their initialization to the source (`LoadSharedEntities` and `ParNCMesh::GetConformingSharedStructures()`). I would just need to find a way how to hijack the path when the topology is changed in `Finalize`.

Reply:
How about now? I added a protected method `UpdatedTopology()` to hijack the case with the change of topology, but it is not fully self-purposed, I believe, and might be useful in the future too.

Reply:
ping @sebastiangrimberg

Reply:
Moved to #4240 .